server/src/main/java/org/opensearch/index/mapper/DenseVectorFieldMapper.java

```diff
@@ -0,0 +1,333 @@
+                throw new IllegalArgumentException(
+                    String.format(Locale.ROOT, "[dimension] value cannot be greater than %d for vector [%s]", MAX_DIMENSION, name)
+                );
+            }
+            if (value <= 0) {
+                throw new IllegalArgumentException(
+                    String.format(Locale.ROOT, "[dimension] value must be greater than 0 for vector [%s]", name)
+                );
+            }
+            return value;
+        }, m -> toType(m).dimension).setSerializer((b, n, v) -> b.field(n, v.intValue()), v -> Integer.toString(v.intValue()));
+
+        private final Parameter<KnnContext> knnContext = new Parameter<>(
+            Names.KNN.getValue(),
+            false,
+            () -> null,
+            (n, c, o) -> KnnContext.parse(o),
+            m -> toType(m).knnContext
+        ).setSerializer(((b, n, v) -> {
+            if (v == null) {
+                return;
+            }
+            b.startObject(n);
+            v.toXContent(b, ToXContent.EMPTY_PARAMS);
+            b.endObject();
+        }), m -> m.getKnnAlgorithmContext().getMethod().name());
+
+        public Builder(String name) {
+            super(name);
+        }
+
+        @Override
+        protected List<Parameter<?>> getParameters() {
+            return List.of(dimension, knnContext, hasDocValues);
+        }
+
+        @Override
+        public DenseVectorFieldMapper build(BuilderContext context) {
+            return new DenseVectorFieldMapper(
+                buildFullName(context),
+                new DenseVectorFieldType(buildFullName(context), dimension.get(), knnContext.get()),
+                multiFieldsBuilder.build(this, context),
+                copyTo.build()
+            );
+        }
+    }
+
+    /**
+     * Type parser for dense_vector field mapper
+     *
+     * @opensearch.internal
+     */
+    public static class TypeParser implements Mapper.TypeParser {
+
+        @Override
+        public Mapper.Builder<?> parse(String name, Map<String, Object> node, ParserContext parserContext) throws MapperParsingException {
+            Builder builder = new DenseVectorFieldMapper.Builder(name);
+            Object dimensionField = node.get(Names.DIMENSION.getValue());
+            String dimension = XContentMapValues.nodeStringValue(dimensionField, null);
+            if (dimension == null) {
+                throw new MapperParsingException(String.format(Locale.ROOT, "[dimension] property must be specified for field [%s]", name));
+            }
+            builder.parse(name, parserContext, node);
+            return builder;
+        }
+    }
+
+    /**
+     * Field type for dense_vector field mapper
+     *
+     * @opensearch.internal
+     */
+    public static class DenseVectorFieldType extends MappedFieldType {
+
+        private final int dimension;
+        private final KnnContext knnContext;
+        private final boolean hasDocValues;
```

Comment:
MappedFieldType [already takes care of `hasDocValues` for you](https://github.com/opensearch-project/OpenSearch/blob/088e019736fc39f7b96ff15731f77df516ca0fb8/server/src/main/java/org/opensearch/index/mapper/MappedFieldType.java#L107). Let's remove this variable and use the built in mechanism.l

Reply:
let me check MappedFieldType logic, initially I had to add custom code because unit tests were failing. Probably some piece was missing.